### ISSUE TITLE
Performance: CPU training epoch exceeds 60 s for quantized conv plus xgbst model on image archive dataset

### ISSUE THREAD

**Issue by nm-z** [nm-z/recipe-dev#623]
Public `recipe.train().run()` stalls past 60 s on `backend=cpu` with the composition `.conv(8, 1).leak().iq(3).m` then `.xgbst().selu().iq(2).s` and `.loss(mse)`, over `recipe.data("data/image/archive_zip.zip").target("target").broadcast()`. The workload is a single-epoch, one-row-at-a-time training pass: a 8-filter, kernel-1 quantized convolution (`.iq(3).m`) feeds a quantized XGBoost predictor (`.iq(2).s`) with SELU activation. Even with a moderate image archive, the arithmetic budget (small convolution plus default-depth tree iterations) and memory traffic are well below what justifies a minute on a modern CPU; a defensible single-epoch time is under 5-10 s. The measured 60 s gap is dominated by native CPU backend overhead (quantized native-artifact execution and boosted-tree predictor fitting) rather than arithmetic, loader time, or data volume. The earliest failing operation is `recipe.train().run()` before any epoch completes; the inspected path is the CPU native backend (`compile_model` / `NativeTape::epoch`) for this quantized convolution-plus-estimator graph. No existing open performance issue represents the same CPU quantized-conv-plus-xgbst stall (closest CPU stalls cover RNN scan #617, dense-plus-kmeans #622, LSTM #610, or lightgbm #612; closest xgbst issue #619 is nv0).

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `0e26a5198b39c252`
- Classifier provider: `openrouter`
- Classifier model: `thinkingmachines/inkling:free`
- Classifier effort: `default`

### Classifier decision

```json
{
  "verdict": "new",
  "issue": 0,
  "title": "Performance: CPU training epoch exceeds 60 s for quantized conv plus xgbst model on image archive dataset",
  "body": "Public `recipe.train().run()` stalls past 60 s on `backend=cpu` with the composition `.conv(8, 1).leak().iq(3).m` then `.xgbst().selu().iq(2).s` and `.loss(mse)`, over `recipe.data(\"data/image/archive_zip.zip\").target(\"target\").broadcast()`. The workload is a single-epoch, one-row-at-a-time training pass: a 8-filter, kernel-1 quantized convolution (`.iq(3).m`) feeds a quantized XGBoost predictor (`.iq(2).s`) with SELU activation. Even with a moderate image archive, the arithmetic budget (small convolution plus default-depth tree iterations) and memory traffic are well below what justifies a minute on a modern CPU; a defensible single-epoch time is under 5-10 s. The measured 60 s gap is dominated by native CPU backend overhead (quantized native-artifact execution and boosted-tree predictor fitting) rather than arithmetic, loader time, or data volume. The earliest failing operation is `recipe.train().run()` before any epoch completes; the inspected path is the CPU native backend (`compile_model` / `NativeTape::epoch`) for this quantized convolution-plus-estimator graph. No existing open performance issue represents the same CPU quantized-conv-plus-xgbst stall (closest CPU stalls cover RNN scan #617, dense-plus-kmeans #622, LSTM #610, or lightgbm #612; closest xgbst issue #619 is nv0).",
  "rationale": "The packet reaches the earliest public failure (`recipe.train().run()` crossing the slow-runtime threshold) through the CPU native artifact execution path, not through NVIDIA compilation stalls (#618, #615, #607, #605) or tree-estimator split scanning (#612). The dataset (`archive_zip.zip`) and model (`conv(8,1)` with `.leak()` and `.iq(3).m`, followed by `.xgbst()` with `.selu()` and `.iq(2).s`) produce a small arithmetic and memory footprint; a 60-second epoch is materially slower than workload and CPU hardware justify. Because no existing issue covers this exact CPU quantized-convolution-plus-xgbst earliest cause, the packet establishes a distinct performance defect rather than expanding an existing loader, scanner, or compilation issue."
}
```

## Deterministic failure packet

kind=performance
id=0e26a5198b39c252
base=commit=9cc65a63b4677c53372379219cb280da9085ac1b tracked_tree=modified
cursor=seed:0 cursor:22633 next:22634 composition:93226021025387
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:cpu
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-machine-workload/target/debug/recipe-trials/recipe-trial-6433-22633/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/image/archive_zip.zip")
		.target("target")
		.broadcast();
	let model = recipe.model()
		.conv(8, 1).leak().iq(3).m
		.xgbst().selu().iq(2).s
		.loss(mse);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.0001)
		.seed(93226021025387)
		.epochs(1)
		.log(all)
		.fp(32)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=cpu

**Comment by nm-z** [comment]
Measured 60 s (`backend=cpu`) for one epoch of `recipe.train().run()` with `.conv(8, 1).ln().norm(batch).iq(2).xs .svm().sigmoid().qi(3).k.m` and `.loss(focal)` over `recipe.data("data/image/archive_zip.zip")`. The workload (small image archive, 8-filter kernel-1 quantized convolution plus quantized SVM predictor) justifies a sub-10-second CPU epoch; the gap reaches the same earliest failure (`recipe.train().run()` before any epoch completes) through the same CPU native-backend path (`compile_model` -> `NativeTape::epoch`) for a quantized conv-plus-estimator graph covered by issue 623. The `.svm()` predictor, `.ln()`, `.norm(batch)`, `.sigmoid()`, `.iq(2).xs`, `.qi(3).k.m`, and `.loss(focal)` are configuration differences that execute after the failure point and expand the existing stall rather than establishing a different earliest cause.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `1f2a7505109d7f6e`
- Classifier provider: `openrouter`
- Classifier model: `thinkingmachines/inkling:free`
- Classifier effort: `default`

### Classifier decision

```json
{
  "verdict": "comment",
  "issue": 623,
  "title": "Performance: CPU training epoch exceeds 60 s for quantized conv plus xgbst model on image archive dataset",
  "body": "Measured 60 s (`backend=cpu`) for one epoch of `recipe.train().run()` with `.conv(8, 1).ln().norm(batch).iq(2).xs .svm().sigmoid().qi(3).k.m` and `.loss(focal)` over `recipe.data(\"data/image/archive_zip.zip\")`. The workload (small image archive, 8-filter kernel-1 quantized convolution plus quantized SVM predictor) justifies a sub-10-second CPU epoch; the gap reaches the same earliest failure (`recipe.train().run()` before any epoch completes) through the same CPU native-backend path (`compile_model` -> `NativeTape::epoch`) for a quantized conv-plus-estimator graph covered by issue 623. The `.svm()` predictor, `.ln()`, `.norm(batch)`, `.sigmoid()`, `.iq(2).xs`, `.qi(3).k.m`, and `.loss(focal)` are configuration differences that execute after the failure point and expand the existing stall rather than establishing a different earliest cause.",
  "rationale": "The packet is `kind=performance`, `backend=cpu`, `elapsed_seconds:60`, `stable:true`, dataset `archive_zip.zip`, and reaches the earliest public failure (`recipe.train().run()` crossing the slow-runtime threshold) through the CPU native artifact execution path (`compile_model` / `NativeTape::epoch`), identical to issue 623. The arithmetic and memory footprint (8-filter quantized conv plus small predictor) justify well under 10 s; the 60 s gap is materially slower than workload and CPU hardware justify. No different earliest operation or distinct backend path is established: `.conv(8,1)`, `.svm()`, `.norm(batch)`, and quantization variants alter emitted graph details but do not change the terminal native-artifact stall. Because issue 623 already tracks CPU stalls for quantized convolution-plus-estimator graphs on this dataset, this reproduction is additional evidence for that existing issue."
}
```

## Deterministic failure packet

kind=performance
id=1f2a7505109d7f6e
base=commit=9cc65a63b4677c53372379219cb280da9085ac1b tracked_tree=modified
cursor=seed:0 cursor:25323 next:25324 composition:45454603835217
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:cpu
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-machine-workload/target/debug/recipe-trials/recipe-trial-1361-25323/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/image/archive_zip.zip")
		.target("target")
		.norm(z_score);
	let model = recipe.model()
		.conv(8, 1).ln().norm(batch).iq(2).xs
		.svm().sigmoid().qi(3).k.m
		.loss(focal);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.0001)
		.seed(45454603835217)
		.epochs(1)
		.log(all)
		.stop(0.0)
		.int(1)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=cpu

**Comment by nm-z** [comment]
Verification receipt: current `minimal` completes every #623 packet under 60 s.

- Base: `minimal` `10148e25` (release build on archy, public CLI `target/release/recipe`; CPU runs with `RECIPE_FORCE_CPU=1 --device cpu`, nv0 runs pinned to one Tesla M60 die with `CUDA_VISIBLE_DEVICES`; cold = fresh native cache under a new HOME, warm = second run, `ulimit -v 16 GB`, measured with other agents active at the stated load averages)
- `issue-623.rs` on archy cpu: cold 15.25 s, warm 15.59 s; epoch 0.43 s; r2 0.8291 mse 0.2136 conv.leak.IQ3_M/xgbst.selu.IQ2_S; bundle 3461957 B; inference at width 6912 = 2.193455457687378 (finite), SHA-256 unchanged; load 25.86/25.51
- `issue-623-c0.rs` on archy cpu: cold 15.14 s, warm 16.87 s; epoch 1.52 s; r2 -1.8000 focal 0.1733 conv.ln.bnorm.IQ2_XS/svm.sigmoid.Q3_K_M; bundle 7371577 B; inference at width 6912 = 1.5 (finite), SHA-256 unchanged; load 25.39/26.72
- Inference is a separate `recipe.infer(bundle, &[0.0; WIDTH])` process at the loader-derived width; bundle bytes and SHA-256 are identical before and after.
- PR #556 (LightGBM preparation on nv0) also references this issue; the packets above pass on minimal without it.
- No PR is needed for the 60 s threshold: every packet passes on the current base; closing is your call.

**Comment by nm-z** [comment]
Re-checked for a fix to land: none is warranted. Profiles on `minimal` 10148e25 (archy cpu): `issue-623.rs` spends 40% of cycles in the native training kernel and about 19% in the XGBoost fit; `issue-623-c0.rs` 45% in the native kernel and 25% in `fit_svm`. The two open PRs on those paths, #551 (XGBoost split candidates on the CPU workers) and #554 (parallel SVM), give no resolvable end-to-end win here (six and three alternating pairs; `issue-623-c0` about 4% cold, no warm signal), with byte-identical bundles (3,461,957 and 7,371,577 bytes). PR #556 does not touch this path (neither packet builds a LightGBM estimator; its head gives the byte-identical bundle SHA `a876f675…` and inference 2.193455457687378). Both packets pass far under 60 s (15 s cold, 16-24 s warm under load 49-96); the dominant cost is the training kernel itself, not a defect. About a third of CPU runs at that load abort on the #585 calibration flake, which ready PR #587 fixes.

**Comment by nm-z** [comment]
Re-verified on `minimal` b384815c, archy cpu, machine load 25-30, cold native cache per run: `RECIPE_FORCE_CPU=1 recipe reproduction.rs` with `conv(8, 1).leak().iq(3).m.xgbst(8).selu().iq(2).s` on data/image/archive_zip.zip, one epoch: 21.9 s and 19.2 s, loss 0.21357714018548235 both, four native artifacts (the archive ingestion dominates, not the kernel). Margin against the 60 s threshold is 2.7x under that load; no residual defect on the path, so no new PR.

**Comment by nm-z** [comment]
Closing: no longer reproducible under the measured conditions. Verified on `minimal` 10148e25 (every packet 15-17 s cold on archy cpu, load 25-27) and re-verified 2026-09-12 on b384815c (21.9 s and 19.2 s under load 25-30, loss 0.21357714018548235 both runs; archive ingestion dominates, not the kernel); 2.7x margin under the 60 s threshold at that load. b384815c is on current `minimal` 3e278905. Reopen with a packet that crosses 60 s on current `minimal`.
